### PR TITLE
:warning:  instrument support for cluster-aware clients

### PR DESCRIPTION
**Pull request by stevekuznetsov**
Signed-off-by: Steve Kuznetsov <skuznets@redhat.com>

Follows up on #53 
Part of https://github.com/kcp-dev/kcp/issues/1943

### CODE REVIEW

**Comment by stevekuznetsov** [comment]
Crap, looks like the linter :vomiting_face: on generics.

File: pkg/cache/listers.go [hunk 1]
```diff
@@ -27,21 +27,29 @@ import (
 	"k8s.io/client-go/tools/cache"
 )
 
-// NewGenericClusterLister creates a new instance for the GenericClusterLister.
-func NewGenericClusterLister(indexer cache.Indexer, resource schema.GroupResource) *GenericClusterLister {
-	return &GenericClusterLister{
+// NewGenericClusterLister creates a new instance for the genericClusterLister.
+func NewGenericClusterLister(indexer cache.Indexer, resource schema.GroupResource) GenericClusterLister {
```
Comment (ncdc) [inline pkg/cache/listers.go:31]: Does this need to return the interface or can we return a `*genericClusterLister`?

File: pkg/client/constructor.go [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+
+import (
+	"net/http"
+	"sync"
+
+	"github.com/kcp-dev/logicalcluster/v2"
+	"k8s.io/client-go/rest"
+)
+
+type ClientConstructor[R any] struct {
```
Comment (ncdc) [inline pkg/client/constructor.go:27]: Feel like calling this `Constructor` so the usage is `client.Constructor`?

File: pkg/client/constructor.go [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+	NewForConfigAndClient func(*rest.Config, *http.Client) (R, error)
+}
+
+type ClientCache[R any] interface {
```
Comment (ncdc) [inline pkg/client/constructor.go:31]: ClientCache -> Cache?

File: pkg/client/constructor.go [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+		clientsByCluster: map[logicalcluster.Name]R{},
+	}
+}
+
+type clientCache[R any] struct {
+	cfg         *rest.Config
+	client      *http.Client
+	constructor *ClientConstructor[R]
+
+	*sync.RWMutex
```
Comment (ncdc) [inline pkg/client/constructor.go:52]: Does this need to be a pointer?

Comment (stevekuznetsov) [reply pkg/client/constructor.go:52]: IIRC it gets passed around, and can't be by value, right?

Comment (ncdc) [reply pkg/client/constructor.go:52]: If it gets passed around, then yes, pointer it is

File: pkg/cache/listers.go (hunk 1 above)
Comment (stevekuznetsov) [reply pkg/cache/listers.go:31]: My linter yells at me for returning unexported types

Comment (ncdc) [reply pkg/cache/listers.go:31]: Should we export the type too?

Comment (stevekuznetsov) [reply pkg/cache/listers.go:31]: Then we need to find a new name for it, etc - I am ambivalent here. Can tackle it next week, will need updates in codegen & client-go as follow-ups.

File: pkg/cache/listers.go [hunk 5]
```diff
@@ -27,21 +27,29 @@ import (
...
+	return &ClusterLister{
 		indexer:  indexer,
 		resource: resource,
 	}
 }
 
-// GenericClusterLister is a lister that supports multiple logical clusters. It can list the entire contents of the backing store, and return individual cache.GenericListers that are scoped to individual logical clusters.
-type GenericClusterLister struct {
+// GenericClusterLister is a lister skin on a generic Indexer
+type GenericClusterLister interface {
```
Comment (ncdc) [inline pkg/cache/listers.go:39]: Does any of our framework/library code need this? Or if it does, can we declare it where it's used?

File: pkg/client/constructor.go [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+
+import (
+	"net/http"
+	"sync"
+
+	"github.com/kcp-dev/logicalcluster/v2"
+	"k8s.io/client-go/rest"
+)
+
+type Constructor[R any] struct {
```
Comment (ncdc) [inline pkg/client/constructor.go:27]: You want to do godoc for all the things before merging?

File: pkg/cache/listers.go (hunk 5 above)
Comment (stevekuznetsov) [reply pkg/cache/listers.go:39]: This (colocating the interface with the impl) is just parallel to how k8s does it, and I think there's more utility to being parallel than doing the "interface at the callsite" thing. Although, in k8s the library is `k8s.io/client-go/tools/cache`. I don't have strong feelings on where this lives but please keep in mind that every day spent on this sort of stuff incurs large rebase cost for the kcp refactor PR.

File: pkg/client/constructor.go (hunk 6 above)
Comment (stevekuznetsov) [reply pkg/client/constructor.go:27]: Sure, can do - but ditto to the rebase cost comment above w.r.t. timelines.

**Review by ncdc — APPROVED** [review]

**Pull request merged**